Adds text embedding and vector search end-to-end example (#2229)
elastic/stack-docs: docs/en/stack/ml/nlp/ml-nlp-text-emb-vector-search-example.asciidoc

    {
      "predicted_value": [
        0.39521875977516174,
        -0.3263707458972931,
        0.26809820532798767,
        0.30127981305122375,
        0.502890408039093,
        ...
      ]
    }
  ]
}
--------------------------------------------------
// NOTCONSOLE
====

The result is the predicted dense vector transformed from the example text.

[discrete]
[[ex-text-emb-data]]
== Load data

The data set 
https://msmarco.blob.core.windows.net/msmarcoranking/msmarco-passagetest2019-top1000.tsv.gz[msmarco-passagetest2019-top1000.tsv] 
is a subset of the MS MACRO Passage Ranking data set used in the testing stage 
of the 2019 TREC Deep Learning Track. It contains 200 queries and for each query 
a list of relevant text passages extracted by a simple IR system. From that 
dataset, all unique passages with their IDs have been extracted and put into a 
https://github.com/elastic/stack-docs/blob/8.5/docs/en/stack/ml/nlp/data/msmarco-passagetest2019-unique.tsv[tsv file], 
totaling 182469 passages. In the following, this file is used as the example 
data set.

Upload the file by using the 
{kibana-ref}/connect-to-elasticsearch.html#upload-data-kibana[Data Visualizer]. 

> Name the first column `id` and the second one `text`. The index name is 

Reviewer: When I tried this with the linked file (msmarco-passagetest2019-top1000.tsv) Data Visualizer found 4 columns, the first 2 were numbers and the last 2 text. Is this the right file?

Author: No, the second tsv file is the one that should be uploaded in the Data Viz (tsv file --> https://github.com/elastic/stack-docs/blob/8.5/docs/en/stack/ml/nlp/data/msmarco-passagetest2019-unique.tsv, this PR contains it). The first link is just a reference to the data set that the tsv file we use has been extracted from. I'll remove the link from the first data set as it's misleading.

Reviewer: Ah ok it worked as expected when I used the second tsv file

Author: I made the link less prominent – and hopefully less misleading, too – in [6a2151f](https://github.com/elastic/stack-docs/pull/2229/commits/6a2151f03b35ad20c49fdd8ac7f74700b9def118).

PUT _ingest/pipeline/text-embeddings
{
  "description": "Text embedding pipeline",
  "processors": [
    {
      "inference": {
        "model_id": "sentence-transformers__msmarco-minilm-l-12-v3",
        "target_field": "text_embedding",
        "field_map": {
          "text": "text_field"
        }
      }
    }
  ],
  "on_failure": [
    {
      "set": {
        "description": "Index document to 'failed-<index>'",
        "field": "_index",
        "value": "failed-{{{_index}}}"
      }
    },
    {
      "set": {
        "description": "Set error message",
        "field": "ingest.failure",
        "value": "{{_ingest.on_failure_message}}"
      }
    }
  ]
}
--------------------------------------------------

The passages are in a field named `text`. The `field_map` maps the text to the 
field `text_field` that the model expects. The `on_failure` handler is set to 
index failures into a different index.

Before ingesting the data through the pipeline, create the mappings of the 
destination index, in particular for the field `text_embedding.predicted_value` 
where the ingest processor stores the embeddings. The used modelproduces 

> embeddings as 384 dimensional vectors, hence you can use the indexed 

Reviewer: ```suggestion
embeddings with 384 dimensions, the `dense_vector` field must be configured 
```

[[ml-nlp-text-emb-vector-search-example]]
= How to deploy a text embedding model and use it with vector search

++++
<titleabbrev>Text embedding and vector search</titleabbrev>
++++
:keywords: {ml-init}, {stack}, {nlp}

You can use these instructions to deploy a 
<<ml-nlp-text-embedding,text embedding>> model in {es}, test the model, and 
add it to an {infer} ingest pipeline. It enables you to generate vector 
representations of text and perform vector similarity search on the generated 
vectors. The model that is used in the example is publicly available on 
https://huggingface.co/[HuggingFace].

> The example uses a public dataset from the 

Reviewer 2: ```suggestion
The example uses a public data set from the 
```

https://microsoft.github.io/msmarco/#ranking[MS MARCO Passage Ranking Task]. It 
consists of real questions from the Microsoft Bing search engine and human 

> generated answers for them. The example works with a sample of this dataset, 

Reviewer 2: This is two words per https://brand.elastic.co/302f66895/p/194a3b-writing-style-guide
```suggestion
generated answers for them. The example works with a sample of this data set, 
```

.**Test the model by using the _infer API**
[%collapsible]
====
You can also evaluate your models by using the 
{ref}/infer-trained-model-deployment.html[_infer API]. In the following request, 
`text_field` is the field name where the model expects to find the input, as 
defined in the model configuration. By default, if the model was uploaded via 
Eland, the input field is `text_field`.

[source,js]
--------------------------------------------------
POST /_ml/trained_models/sentence-transformers__msmarco-minilm-l-12-v3/_infer
{
  "docs": {
    "text_field": "How is the weather in Jamaica?"
  }
}
--------------------------------------------------

The API returns a response similar to the following:

[source,js]
--------------------------------------------------
{
  "inference_results": [
    {
      "predicted_value": [
        0.39521875977516174,
        -0.3263707458972931,
        0.26809820532798767,
        0.30127981305122375,
        0.502890408039093,
        ...
      ]
    }
  ]
}
--------------------------------------------------
// NOTCONSOLE
====

The result is the predicted dense vector transformed from the example text.

[discrete]
[[ex-text-emb-data]]

> == Load data

Reviewer 2: > == Load data

Maybe it's just me, but I found when I got to this step I didn't remember *why* I'd be loading this data (and what its purpose was).  A reminder that this is data that you'll use in your ingest pipeline might help those of us who are losing the path.

> The data set msmarco-passagetest2019-top1000.tsv (you can find it 

Reviewer 2: > The data set msmarco-passagetest2019-top1000.tsv (you can find it 

It seems like there are links to two different tsv files in this paragraph, which makes me a little uncertain which one I'm supposed to use. If we're *not* expecting users to download this first tsv file, could we link to an info page or somehow make it clearer that we're linking to this only for information purposes?

https://msmarco.blob.core.windows.net/msmarcoranking/msmarco-passagetest2019-top1000.tsv.gz[here]) 
is a subset of the MS MACRO Passage Ranking data set used in the testing stage 
of the 2019 TREC Deep Learning Track. It contains 200 queries and for each query 

> a list of relevant text passages extracted by a simple IR system. From that 

Reviewer 2: > a list of relevant text passages extracted by a simple IR system. From that 

What is the meaning of IR? Information retrieval? Since this is the first use on this page, I think it should be defined.

> dataset, all unique passages with their IDs have been extracted and put into a 

Reviewer 2: ```suggestion
data set, all unique passages with their IDs have been extracted and put into a 
```

https://github.com/elastic/stack-docs/blob/8.5/docs/en/stack/ml/nlp/data/msmarco-passagetest2019-unique.tsv[tsv file], 
totaling 182469 passages. In the following, this file is used as the example 
data set.

Upload the file by using the 
{kibana-ref}/connect-to-elasticsearch.html#upload-data-kibana[Data Visualizer]. 
Name the first column `id` and the second one `text`. The index name is 
`collection`. After the upload is done, you can see an index named `collection` 
with 182469 documents.

[role="screenshot"]
image::images/ml-nlp-text-emb-data.png[Importing the data]

[discrete]
[[ex-text-emb-ingest]]
== Add the text embedding model to an {infer} ingest pipeline

Process the initial data with an 
{ref}/inference-processor.html[{infer} processor]. It adds an embedding for each 
passage. For this, create a text embedding ingest pipeline and then reindex the 
initial data with this pipeline.

Now create an ingest pipeline either in the 

> {ml-docs}/ml-nlp-inference.html#ml-nlp-inference-processor[Stack management UI] 

Reviewer 2: ```suggestion
{ml-docs}/ml-nlp-inference.html#ml-nlp-inference-processor[{stack-manage-app} UI] 
```

PUT _ingest/pipeline/text-embeddings
{
  "description": "Text embedding pipeline",
  "processors": [
    {
      "inference": {
        "model_id": "sentence-transformers__msmarco-minilm-l-12-v3",
        "target_field": "text_embedding",
        "field_map": {
          "text": "text_field"
        }
      }
    }
  ],
  "on_failure": [
    {
      "set": {
        "description": "Index document to 'failed-<index>'",
        "field": "_index",
        "value": "failed-{{{_index}}}"
      }
    },
    {
      "set": {
        "description": "Set error message",
        "field": "ingest.failure",
        "value": "{{_ingest.on_failure_message}}"
      }
    }
  ]
}
--------------------------------------------------

The passages are in a field named `text`. The `field_map` maps the text to the 
field `text_field` that the model expects. The `on_failure` handler is set to 
index failures into a different index.

Before ingesting the data through the pipeline, create the mappings of the 
destination index, in particular for the field `text_embedding.predicted_value` 
where the ingest processor stores the embeddings. The msmarco-MiniLM-L-12-v3 model produces 

> embeddings with 384 dimensions, the `dense_vector` field must be configured 

Reviewer 2: These could alternatively be split into two sentences IMO
```suggestion
embeddings with 384 dimensions; the `dense_vector` field must be configured 
```

    "pipeline": "text-embeddings"
  }
}
--------------------------------------------------

The API call returns a task ID that can be used to monitor the progress:

[source,js]
--------------------------------------------------
GET _tasks/<task_id>
--------------------------------------------------

You can also open the model stat UI to follow the progress.

[role="screenshot"]
image::images/ml-nlp-text-emb-reindex.png[Model status UI]

After the reindexing is finished, the documents in the new index contain the 
{infer} results – the vector embeddings. 

[discrete]
[[ex-text-emb-vect-search]]
== Vector similarity search

To perform vector similarity search, you need to obtain the text embedding of a 
text. This example uses the "How is the weather in Jamaica?" query as the input 
text. The {ref}/infer-trained-model-deployment.html[_infer API] gives you the 
embedding of this query as a dense vector:

[source,js]
--------------------------------------------------
POST /_ml/trained_models/sentence-transformers__msmarco-minilm-l-12-v3/_infer
{
  "docs": {
    "text_field": "How is the weather in Jamaica?"
  }
}
--------------------------------------------------

> You can use the dense vector in the `collection-with-embeddings` index by doing 

Reviewer 2: I think we're using the results from the previous API as the query_vector field in the following API. Can we make this a little clearer? For example:
```suggestion
You can use the resulting dense vector in the `query_vector` of a 
```

{ref}/knn-search.html[kNN search]:

[source,js]
--------------------------------------------------
GET collection-with-embeddings/_search
{
  "knn": {
    "field": "text_embedding.predicted_value",
    "query_vector": [
     0.39521875977516174,
        -0.3263707458972931,
        0.26809820532798767,
        0.30127981305122375,
     (...)
    ],
    "k": 10,
    "num_candidates": 100
  },
  "_source": [
    "id",
    "text"
  ]
}
--------------------------------------------------

> As a result, you receive the top 10 closest to the query documents from the 

Reviewer 2: This feels like a word is missing:
> As a result, you receive the top 10 closest xxx to the query documents from the 

Is it documents? Passages? Strings?

Author: The expression "closest to the query" works as an adjective that refers to the noun "documents". I rephrased it to the following: "As a result, you receive the top 10 documents that are the closest in meaning to the query..."
